### PR TITLE
chore(auth): loading github module optionally

### LINKED ISSUE #107: Make GitHub OAuth optional
## The problem

In many scenarios, we want the GitHub OAuth to be optional. Currently its not possible since the application will crash if the github specific oauth keys are not mentioned

## Solution

Make the GitHub OAuth provider check for `GITHUB_CLIENT_ID`, `GITHUB_CLIENT_SECRET` and `GITHUB_CALLBACK_URL`. If any one of them is underfined, we don't want the application to init GitHub OAuth.

Additionally, we would also like the controller in charge of doing github login to send a message that Github log in is not enabled in this environment with a `BadRequestException`

### PR DESCRIPTION
**Pull request by HarshPatel5940**
## Description
This pull request refactors the Github authentication module to load the Github module optionally. It also adds a new service called `GithubEnvService` to handle Github environment configurations. The `GithubStrategy` class now accepts the Github credentials as constructor parameters. Additionally, the `AuthController` now checks if Github login is enabled before redirecting to the Github OAuth endpoint.

Fixes #107 

## Mentions
@rajdip-b 

image

## Developer's checklist

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-check on my work

- [x] I have followed the [coding guidelines](https://google.github.io/styleguide/jsguide.html)
- [x] My changes in code generate no new warnings
- [ ] My changes are breaking another fix/feature of the project
- [ ] I have added test cases to show that my feature works
- [x] I have added relevant screenshots in my PR
- [x] There are no UI/UX issues

### Documentation Update
- [ ] This PR requires an update to the documentation at [docs.keyshade.xyz](https://docs.keyshade.xyz)
- [ ] I have made the necessary updates to the documentation, or no documentation changes are required.

___

## **Type**
enhancement, documentation

___

## **Description**
- Refactored GitHub authentication to use `GithubOAuthStratergyFactory` for dynamic strategy creation based on environment configuration.
- Introduced `ConfigModule` to manage OAuth strategies, including GitHub.
- Enhanced `AuthController` to check if GitHub authentication is enabled before proceeding with OAuth flow.
- Added new tests and updated existing ones to accommodate changes in authentication flow.
- Documentation should be updated to reflect changes in GitHub OAuth setup and configuration.

___

## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>auth.module.ts</strong><dd><code>Refactor AuthModule to Use GithubOAuthStrategyFactory</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/api/src/auth/auth.module.ts
<li>Refactored to use <code>GithubOAuthStratergyFactory</code> for creating <br><code>GithubStrategy</code>.<br> <li> Updated providers to include <code>GithubOAuthStratergyFactory</code> and <br>dynamically create <code>GithubStrategy</code>.<br>

</details>
    

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/112/files#diff-75dfb447c1d9cebb6fcd115c5aa5966231dbcc6eff2d66241ad3de04be89078b">+13/-2</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>auth.controller.ts</strong><dd><code>Enhance GitHub OAuth Login Flow with Enablement Check</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/api/src/auth/controller/auth.controller.ts
<li>Added <code>GithubOAuthStratergyFactory</code> to constructor for checking GitHub <br>OAuth enablement.<br> <li> Modified <code>githubOAuthLogin</code> to check if GitHub Auth is enabled before <br>proceeding.<br>

</details>
    

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/112/files#diff-ca5c58065b10468a07314a5a9224336ac37df84a4a0e769babb2ed040cf8c82a">+16/-8</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>config.module.ts</strong><dd><code>Introduce ConfigModule for GitHub OAuth Strategy</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/api/src/config/config.module.ts
<li>Introduced <code>ConfigModule</code> to provide <code>GithubStrategy</code> and <br><code>GithubOAuthStratergyFactory</code>.<br>

</details>
    

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/112/files#diff-1b5afd9670c33b8040cbfd53732ba6e5baaf663eaa8716251081bb7f08c44728">+9/-0</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>github-stratergy.factory.ts</strong><dd><code>Implement GithubOAuthStrategyFactory for Dynamic Strategy Creation</code></dd></summary>
<hr>

apps/api/src/config/factory/github-stratergy.factory.ts
<li>Implemented <code>GithubOAuthStratergyFactory</code> for creating GitHub strategy <br>based on environment config.<br> <li> Added methods to check if GitHub config exists and to create OAuth <br>strategy.<br>

</details>
    

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/112/files#diff-ab7dd256df1e577405b6ae36808dabf77f58133bf4f6a33fb0b2e0ebf8a895be">+54/-0</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>oauth-stratergy.factory.ts</strong><dd><code>Define OAuthStrategyFactory Interface</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/api/src/config/factory/oauth-stratergy.factory.ts
<li>Defined <code>OAuthStratergyFactory</code> interface for OAuth strategy creation.<br>

</details>
    

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/112/files#diff-5fdcae23837e7822b8919f9b2131af6cbe9790d10ce6f0c4317cac483a9af481">+5/-0</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>github.stratergy.ts</strong><dd><code>Refactor GitHub Strategy to Use Constructor Parameters</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/api/src/config/oauth-stratergy/github.stratergy.ts
<li>Refactored to accept credentials via constructor instead of <br><code>ConfigService</code>.<br> <li> Updated <code>validate</code> method signature for clarity.<br>

</details>
    

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/112/files#diff-dc5f8921ee77e01e486498f611a9fbaece323901d2903e885cc63fa537b66f3d">+6/-6</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>auth.controller.spec.ts</strong><dd><code>Update AuthController Tests with New Dependencies</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/api/src/auth/controller/auth.controller.spec.ts
<li>Added <code>GithubOAuthStratergyFactory</code> and <code>ConfigService</code> to providers in <br>tests.<br>

</details>
    

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/112/files#diff-36930809962c9b0ad810d66fc3d4711e1e6c74be800231f46e9c3df907d4a1c4">+4/-0</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    
</table></td></tr></tr></tbody></table><hr>

<details> <summary><strong>✨ Usage guide:</strong></summary><hr> 

**Overview:**
The `describe` tool scans the PR code changes, and generates a description for the PR - title, type, summary, walkthrough and labels. The tool can be triggered [automatically](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#github-app-automatic-tools) every time a new PR is opened, or can be invoked manually by commenting on a PR.

When commenting, to edit [configurations](https://github.com/Codium-ai/pr-agent/blob/main/pr_agent/settings/configuration.toml#L46) related to the describe tool (`pr_description` section), use the following template:
```
/describe --pr_description.some_config1=... --pr_description.some_config2=...
```
With a [configuration file](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#working-with-github-app), use the following template:
```
[pr_description]
some_config1=...
some_config2=...
```

<table><tr><td><details> <summary><strong> Enabling\disabling automation </strong></summary><hr>

- When you first install the app, the [default mode](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#github-app-automatic-tools) for the describe tool is:
```
pr_commands = ["/describe --pr_description.add_original_user_description=true" 
                         "--pr_description.keep_original_user_title=true", ...]
```
meaning the `describe` tool will run automatically on every PR, will keep the original title, and will add the original user description above the generated description. 

- Markers are an alternative way to control the generated description, to give maximal control to the user. If you set:
```
pr_commands = ["/describe --pr_description.use_description_markers=true", ...]
```
the tool will replace every marker of the form `pr_agent:marker_name` in the PR description with the relevant content, where `marker_name` is one of the following:
  - `type`: the PR type.
  - `summary`: the PR summary.
  - `walkthrough`: the PR walkthrough.

Note that when markers are enabled, if the original PR description does not contain any markers, the tool will not alter the description at all.
        

</details></td></tr>

<tr><td><details> <summary><strong> Custom labels </strong></summary><hr>

The default labels of the `describe` tool are quite generic: [`Bug fix`, `Tests`, `Enhancement`, `Documentation`, `Other`].

If you specify [custom labels](https://github.com/Codium-ai/pr-agent/blob/main/docs/DESCRIBE.md#handle-custom-labels-from-the-repos-labels-page-gem) in the repo's labels page or via configuration file, you can get tailored labels for your use cases.
Examples for custom labels:
- `Main topic:performance` - pr_agent:The main topic of this PR is performance
- `New endpoint` - pr_agent:A new endpoint was added in this PR
- `SQL query` - pr_agent:A new SQL query was added in this PR
- `Dockerfile changes` - pr_agent:The PR contains changes in the Dockerfile
- ...

The list above is eclectic, and aims to give an idea of different possibilities. Define custom labels that are relevant for your repo and use cases.
Note that Labels are not mutually exclusive, so you can add multiple label categories.
Make sure to provide proper title, and a detailed and well-phrased description for each label, so the tool will know when to suggest it.        

</details></td></tr>

<tr><td><details> <summary><strong> Inline File Walkthrough 💎</strong></summary><hr>

For enhanced user experience, the `describe` tool can add file summaries directly to the "Files changed" tab in the PR page.
This will enable you to quickly understand the changes in each file, while reviewing the code changes (diffs).

To enable inline file summary, set `pr_description.inline_file_summary` in the configuration file, possible values are:
- `'table'`: File changes walkthrough table will be displayed on the top of the "Files changed" tab, in addition to the "Conversation" tab.
- `true`: A collapsable file comment with changes title and a changes summary for each file in the PR.
- `false` (default): File changes walkthrough will be added only to the "Conversation" tab.
<tr><td><details> <summary><strong> Utilizing extra instructions</strong></summary><hr>

The `describe` tool can be configured with extra instructions, to guide the model to a feedback tailored to the needs of your project.

Be specific, clear, and concise in the instructions. With extra instructions, you are the prompter. Notice that the general structure of the description is fixed, and cannot be changed. Extra instructions can change the content or style of each sub-section of the PR description.

Examples for extra instructions:
```
[pr_description] 
extra_instructions="""
- The PR title should be in the format: '<PR type>: <title>'
- The title should be short and concise (up to 10 words)
- ...
"""
```
Use triple quotes to write multi-line instructions. Use bullet points to make the instructions more readable.

</details></td></tr>

<tr><td><details> <summary><strong> More PR-Agent commands</strong></summary><hr> 

> To invoke the PR-Agent, add a comment using one of the following commands:  
> - **/review**: Request a review of your Pull Request.   
> - **/describe**: Update the PR title and description based on the contents of the PR.   
> - **/improve [--extended]**: Suggest code improvements. Extended mode provides a higher quality feedback.   
> - **/ask \<QUESTION\>**: Ask a question about the PR.   
> - **/update_changelog**: Update the changelog based on the PR's contents.   
> - **/add_docs** 💎: Generate docstring for new components introduced in the PR.   
> - **/generate_labels** 💎: Generate labels for the PR based on the PR's contents.   
> - **/analyze** 💎: Automatically analyzes the PR, and presents changes walkthrough for each component.   

>See the [tools guide](https://github.com/Codium-ai/pr-agent/blob/main/docs/TOOLS_GUIDE.md) for more details.
>To list the possible configuration parameters, add a **/config** comment.   
 

</details></td></tr>

</table>

See the [describe usage](https://github.com/Codium-ai/pr-agent/blob/main/docs/DESCRIBE.md) page for a comprehensive guide on using this tool.

</details>

### CODE REVIEW

**Review by rajdip-b — CHANGES_REQUESTED** [review]
Alright I went through the changes, and the logic looks fine to me. (Thanks for the screenshot ❤️ )

Here's a list of things I would like you to do to make this complete:
- [ ] Create another module named `config`: `cd apps/api && nest g module config`
- [ ] Create a folder named `oauth-stratergy` and move the `github.stratergy.ts` file to this folder
- [ ] Create a folder inside that module named `factory`
- [ ] Add an interface inside it named `OAuthStratergyFactory (oauth-stratergy.factory.ts)` with just one function in it -> `<T extends PassportStratergy> createOAuthStratergy(): T`
- [ ] Add a service class named `GitHubOAuthStratergyFactory (github-stratergy.factory.ts)` in the `factory` folder, implement the `OAuthStratergy` interface, and combine the code that you wrote in `GithubEnvService`. This new factory class will have two roles: a private function to check if github is enabled or not, a public function `createOAuthStratergy` that will do the work you are currently doing in the provider section of the `auth.module.ts` file.
- [ ] Export these files from the config folder using the `exports` array.
- [ ] Import them in the `auth.module.ts` and fill in the gaps

If you can spot, I'm motivating you to use factory pattern for this one. This is for a bigger cause:
- We will have multiple auth providers in future (even metamask wallet)
- We will have a couple of services (sentry for example) that we would want to configure in a similar manner. 

Your PR here would lay down the fundamentals of a very big architecture of our repository ❤️

File: apps/api/src/auth/auth.module.ts [hunk 1]
```diff
@@ -18,7 +18,27 @@ import { GithubStrategy } from './github.stratergy'
...
+      useFactory: (githubEnvService: GithubEnvService) => {
+        if (githubEnvService.isGithubEnabled()) {
+          const creds = githubEnvService.getGithubCredentials()
+          return new GithubStrategy(
+            creds.clientID,
+            creds.clientSecret,
+            creds.callbackURL
+          )
+        } else {
+          Logger.warn('Github Login Is Not Enabled In This Environment')
```
[the remark this answers is not in the document]
Comment (HarshPatel5940) [reply apps/api/src/auth/auth.module.ts:35]: Got it.

**Comment by HarshPatel5940** [comment]
Thank you @rajdip-b for your time writing these detailed steps, it will immensely help me. I will make the necessary changes as soon as time permits and let you know here.

**Comment by rajdip-b** [comment]
Awesome man @HarshPatel5940 !! Really thanks for doing all this.

**Comment by HarshPatel5940** [comment]
> Awesome man @HarshPatel5940 !! Really thanks for doing all this.

Thanks to you as well for the detailed review!

**Comment by rajdip-b** [comment]
@HarshPatel5940 I'm working on another branch that aims to fix some issues with the CI and also add a functionality to the api-keys. Once that's done, I'll update your branch and make the review again!

**Pull request merged**